src/libsystemd/sd-device/sd-device.c

```diff
@@ -419,6 +419,31 @@ _public_ int sd_device_new_from_stat_rdev(sd_device **ret, const struct stat *st
         return sd_device_new_from_devnum(ret, type, st->st_rdev);
 }
 
+_public_ int sd_device_new_from_devname(sd_device **ret, const char *devname) {
+        struct stat st;
+
+        assert_return(ret, -EINVAL);
+        assert_return(devname, -EINVAL);
+
+        if (!path_startswith(devname, "/dev"))
+                return -EINVAL;
+
+        if (stat(devname, &st) < 0)
+                return errno == ENOENT ? -ENODEV : -errno;
+
+        return sd_device_new_from_stat_rdev(ret, &st);
+}
+
+_public_ int sd_device_new_from_syspath_or_devname(sd_device **ret, const char *path) {
```

Comment:
maybe call this `sd_device_new_from_path()`? that leaves the door open that we can add more kinds of paths? dunno?